docs/src/main/asciidoc/native-reference.adoc

```diff
@@ -153,6 +153,299 @@ $ quarkus-project-0.1-SNAPSHOT-runner -XX:PrintFlags=
+----
+$ ps -o pid,rss,command -p $(pidof code-with-quarkus-1.0.0-SNAPSHOT-runner)
+      PID   RSS COMMAND
+     1915 35472 ./target/code-with-quarkus-1.0.0-SNAPSHOT-runner -Xmx128m
+----
+
+How come this Quarkus native executable consumes ~35MB RSS on startup?
+To get an understanding of this number, this section will use `perf` to trace calls to `syscalls:sys_enter_mmap`.
+This system call is particularly interesting for GraalVM based native applications because of how it allocates heap.
+This is allocated using either aligned or unaligned heap chunks.
+All non-array objects get allocated in thread local aligned chunks.
+Each of these are 1MB in size by default.
+For arrays, if they are bigger than 1/8 of the aligned chunk size,
+they will be allocated in unaligned heap chunks which have a size dependant on the object itself.
+The very first time a thread allocates an object or small array,
+it will request an aligned heap chunk that it will use exclusively until it has run out of space in that chunk,
+in which case it will request another aligned heap chunk.
+So by tracing these system calls,
+the code paths that end up requesting new aligned or unaligned heap chunks will be recorded.
+Next, run the Quarkus native executable through `perf record` tracing the `mmap` system call:
+
+[source,bash]
+----
+$ sudo perf record -e syscalls:sys_enter_mmap --call-graph dwarf -a -- target/code-with-quarkus-1.0.0-SNAPSHOT-runner
+----
+
+[TIP]
+====
+The size of the aligned heap chunks can be changed during native build time.
+A custom value (in number of bytes) can be passed via the `-H:AlignedHeapChunkSize` flag.
+One caveat is that this value needs to be multiple of 4096 (4KB).
+====
+
+Once the startup completes, stop the process and generate the stacks:
+
+[source,bash]
+----
+$ perf script > out.stacks
+----
+
+As a final step, generate a
+https://github.com/brendangregg/FlameGraph[flamegraph]
+with the generated stacks:
+
+[source,bash]
+----
+$ export FG_HOME=...
+$ ${FG_HOME}/stackcollapse-perf.pl < out.stacks | ${FG_HOME}/flamegraph.pl \
+    --color=mem --title="mmap Flame Graph" --countname="calls" > out.svg
+----
+
+The flamegraph should look something close to this:
+
+image::mmap.svg[mmap flamegraph]
+
+There are several things of interest to notice there:
+
+* The stack traces that contain method calls to `com.oracle.svm.core.genscavenge.ThreadLocalAllocation` are related to aligned or unaligned heap chunk allocations explained above.
+  As explained above,
+  for the majority of allocations these chunks will be 1MB by default,
+  so they're interesting because each allocated chunk has a considerable effect on the RSS consumption.
+* Of the thread allocation stacks,
+  the ones under `start_thread` are particularly revealing.
+  The machine on which this test was run has 32 cores,
+  and the default Vert.x/Netty behaviour is to create as many threads as cores.
+  In this stack trace you can observe 2x36 calls to `mmap` system call.
```

Comment:
Good question. Looking at the number of threads is actually lower, it's 18:

```
$ ps -e -T | grep -c $(pidof code-with-quarkus-1.0.0-SNAPSHOT-runner)
18
```

Something has changed that I'm unaware, looking...

Also, it would seem those stacktraces are doubled, it would seem each thread might call twice? 2x2x18. I will investigate why this would be the case.

Reply:
To be more precise, Vert.x/Netty used to create as many as event loop threads as cores, but now in a machine of 32 cores it creates 10 event loop threads. Amongst the other 8 threads there are 2 Vert.x/Netty IO threads:

```
$ ps -e -T | grep $(pidof code-with-quarkus-1.0.0-SNAPSHOT-runner)
   2320    2320 pts/0    00:00:00 code-with-quark
   2320    2321 pts/0    00:00:00 ference Handler
   2320    2322 pts/0    00:00:00 gnal Dispatcher
   2320    2324 pts/0    00:00:00 ecutor-thread-0
   2320    2325 pts/0    00:00:00 -thread-checker
   2320    2326 pts/0    00:00:00 ntloop-thread-0
   2320    2327 pts/0    00:00:00 ntloop-thread-1
   2320    2328 pts/0    00:00:00 ntloop-thread-2
   2320    2329 pts/0    00:00:00 ntloop-thread-3
   2320    2330 pts/0    00:00:00 ntloop-thread-4
   2320    2331 pts/0    00:00:00 ntloop-thread-5
   2320    2332 pts/0    00:00:00 ntloop-thread-6
   2320    2333 pts/0    00:00:00 ntloop-thread-7
   2320    2334 pts/0    00:00:00 ntloop-thread-8
   2320    2335 pts/0    00:00:00 ntloop-thread-9
   2320    2336 pts/0    00:00:00 tloop-thread-10
   2320    2337 pts/0    00:00:00 tloop-thread-11
   2320    2338 pts/0    00:00:00 ceptor-thread-0
```

Reply:
Quarkus code has evolved to be more ergonomic, so it also takes memory available to decide how many event loop threads to start. On my system with Xmx128m it'll create 12 event loop threads.

Reply:
Going back to the maths. I'll rewrite the section in a bit with my findings. This is what is happening:

When you pass -Xmx128m, it trims down the number of event loop threads to 12. Aside from that there are 6 other threads, so 18 threads in total. Now, all of those threads have their name truncated to 16 chars (not 15). Technically, each of those should result in 1 mmap call reserve and 1 mmap commit, making a total of 36 but you see a total of 72. I think what's happening is this:

When you record `syscalls:sys_enter_mmap`, it tracks calls to `__GI___mmap64` and looking at this with `gdb`, you can see that each `mmap` call results in two calls to `__GI___mmap64`. It seems there's some delegation happening:

```
Thread 2 "code-with-quark" hit Breakpoint 1, __GI___mmap64 (offset=0, fd=-1, flags=16418, prot=0, len=2097152, addr=0x0) at ../sysdeps/unix/sysv/linux/mmap64.c:58
58	  return (void *) MMAP_CALL (mmap, addr, len, prot, flags, fd, offset);
(gdb) bt
#0  __GI___mmap64 (offset=0, fd=-1, flags=16418, prot=0, len=2097152, addr=0x0) at ../sysdeps/unix/sysv/linux/mmap64.c:58
#1  __GI___mmap64 (addr=0x0, len=2097152, prot=0, flags=16418, fd=-1, offset=0) at ../sysdeps/unix/sysv/linux/mmap64.c:46
#2  0x00000000004f4033 in com.oracle.svm.core.posix.headers.Mman$NoTransitions::mmap (__0=<optimized out>, __1=<optimized out>, __2=<optimized out>, __3=<optimized out>, __4=<optimized out>, __5=<optimized out>)
#3  0x00000000004f194e in com.oracle.svm.core.posix.PosixVirtualMemoryProvider::reserve (this=0x7ffff7691220, nbytes=0x100000, alignment=0x100000, executable=false) at com/oracle/svm/core/posix/PosixVirtualMemoryProvider.java:126
#4  0x00000000004ef3b3 in com.oracle.svm.core.os.AbstractCommittedMemoryProvider::allocate (this=0x7ffff7658cb0, size=0x100000, alignment=0x100000, executable=false) at com/oracle/svm/core/os/AbstractCommittedMemoryProvider.java:124
#5  0x0000000000482f40 in com.oracle.svm.core.os.AbstractCommittedMemoryProvider::allocateAlignedChunk (this=0x7ffff7658cb0, nbytes=0x100000, alignment=0x100000) at com/oracle/svm/core/os/AbstractCommittedMemoryProvider.java:107
#6  com.oracle.svm.core.genscavenge.HeapChunkProvider::produceAlignedChunk (this=0x7ffff7444398) at com/oracle/svm/core/genscavenge/HeapChunkProvider.java:112
#7  0x0000000000489485 in com.oracle.svm.core.genscavenge.ThreadLocalAllocation::slowPathNewArrayLikeObject0 (hub=0x7ffff6ff6110, length=15, size=0x20, podReferenceMap=0x7ffff6700000) at com/oracle/svm/core/genscavenge/ThreadLocalAllocation.java:306
#8  0x0000000000489165 in com.oracle.svm.core.genscavenge.ThreadLocalAllocation::slowPathNewArrayLikeObject (objectHeader=0x8f6110 <io.smallrye.common.expression.ExpressionNode::toString+160>, length=15, podReferenceMap=0x7ffff6700000) at com/oracle/svm/core/genscavenge/ThreadLocalAllocation.java:279
#9  0x0000000000489066 in com.oracle.svm.core.genscavenge.ThreadLocalAllocation::slowPathNewArray (objectHeader=0x8f6110 <io.smallrye.common.expression.ExpressionNode::toString+160>, length=15) at com/oracle/svm/core/genscavenge/ThreadLocalAllocation.java:242
#10 0x0000000000d202a1 in java.util.Arrays::copyOfRange (original=0x7ffff6a33410, from=2, to=17) at java/util/Arrays.java:3819
#11 0x0000000000acf8e6 in java.lang.StringLatin1::newString (val=0x7ffff6a33410, index=2, len=15) at java/lang/StringLatin1.java:769
#12 0x0000000000acac59 in java.lang.String::substring (this=0x7ffff6dc0d48, beginIndex=2, endIndex=17) at java/lang/String.java:2712
#13 0x0000000000acaba2 in java.lang.String::substring (this=0x7ffff6dc0d48, beginIndex=2) at java/lang/String.java:2680
#14 0x00000000004f96cd in com.oracle.svm.core.posix.thread.PosixPlatformThreads::setNativeName (this=0x7ffff7658d10, thread=0x7ffff723fb30, name=0x7ffff6dc0d48) at com/oracle/svm/core/posix/thread/PosixPlatformThreads.java:163
#15 0x00000000004f9285 in com.oracle.svm.core.posix.thread.PosixPlatformThreads::beforeThreadRun (this=0x7ffff7658d10, thread=0x7ffff723fb30) at com/oracle/svm/core/posix/thread/PosixPlatformThreads.java:212
#16 0x00000000005237a2 in com.oracle.svm.core.thread.PlatformThreads::threadStartRoutine (threadHandle=0x1) at com/oracle/svm/core/thread/PlatformThreads.java:760
#17 0x00000000004f9627 in com.oracle.svm.core.posix.thread.PosixPlatformThreads::pthreadStartRoutine (data=0x2a06e20) at com/oracle/svm/core/posix/thread/PosixPlatformThreads.java:203
#18 0x0000000000462ab0 in com.oracle.svm.core.code.IsolateEnterStub::PosixPlatformThreads_pthreadStartRoutine_38d96cbc1a188a6051c29be1299afe681d67942e (__0=<optimized out>) at com/oracle/svm/core/code/IsolateEnterStub.java:1
#19 0x00007ffff7e4714d in start_thread (arg=<optimized out>) at pthread_create.c:442
#20 0x00007ffff7ec8950 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```

The `mmap64.c` source code for my env can be found [here](https://gist.github.com/galderz/9c178691918a5442e36ec45ae0fb0319#file-gistfile2-txt-L58).

Reply:
Section reworked and pushed.